src/main/kotlin/org/elm/ide/intentions/InlineDebugIntention.kt

```diff
@@ -0,0 +1,95 @@
+ * (since a composition produces a function, which can't be sensibly logged).
+ *
+ * Ideally, we would support inserting Debug.log statements inside a pipeline/composition,
+ * to allow users to print the data getting piped along.
+ */
+class InlineDebugIntention : ElmAtCaretIntentionActionBase<InlineDebugIntention.Context>() {
+    data class Context(val valueToDebug: ElmPsiElement)
+
+    override fun getText() = "Log this value to console"
+    override fun getFamilyName() = text
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
+        element.ancestors.forEach {
+            when (it) {
+                is ElmValueExpr -> {
+                    val functionCallParent = it.parentOfType<ElmFunctionCallExpr>()
+                    val binOpParent = it.parentOfType<ElmBinOpExpr>()
+                    return when {
+                        functionCallParent != null && functionCallParent.target == it -> {
+                            Context(functionCallParent)
+                        }
+                        binOpParent != null -> {
+                            if (isFunComposition(binOpParent)) {
+                                null
+                            } else {
+                                Context(binOpParent)
+                            }
+                        }
+                        else -> {
+                            Context(it)
+                        }
+                    }
+                }
+                is ElmCaseOfExpr -> {
+                    return Context(it)
+                }
+                is ElmAtomTag -> {
+                    return Context(it)
+                }
+                is ElmBinOpExpr -> {
+                    return if (isFunComposition(it)) {
+                        null
+                    } else {
+                        Context(it)
+                    }
+                }
+            }
+        }
+        return null
+    }
+
+    private fun isFunComposition(opExpr: ElmBinOpExpr): Boolean =
+            opExpr.parts
+                    .filterIsInstance<ElmOperator>()
+                    .map { it.referenceName }
+                    .any { it == "<<" || it == ">>" }
+
+
+    override fun invoke(project: Project, editor: Editor, context: Context) {
+        val factory = ElmPsiFactory(project)
+        val valueExpr = context.valueToDebug
+        val itemsText = valueExpr.text
+        val textLines = itemsText.lines()
+
+        val debuggedText =
+                if (textLines.size > 1) {
+                    val baseIndent = editor.getIndent(valueExpr.startOffset)
+                    val indentation = "    "
+                    val indentedLines = textLines.mapIndexed { index, line -> baseIndent + indentation.repeat(index) + line }
+                    """Debug.log
+$baseIndent    "${textLines.first()} ..."
+$baseIndent    (
+$baseIndent    ${indentedLines.joinToString("\n")}
+$baseIndent    )"""
```

Comment:
I'm unsure about how to best specify the indentation here

Reply:
you and me both. It's kinda terrible. Doing any kind of code-gen that involves multiple lines is a pain.